File: Library/activiser.Library.Signature.Desktop/DatePickerPopup.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace activiser.Library
{
    public partial class DatePickerPopup : Form
    {
        public DatePickerPopup()
        {
            InitializeComponent();
        }

        private Size borderSize = new Size(2, 2);
        private DateTime _value;
        private Control _anchorControl;

        internal DateTime Value
        {
            get
            {
                return _value;
            }
            set
            {
                _value = value;
                this.Calendar.SelectionStart = value;
            }
        }

        internal Control AnchorControl
        {
            get
            {
                return _anchorControl;
            }
            set
            {
                _anchorControl = value;
                if (value != null) Utilities.LocateDropDown(this, this._anchorControl, true);
            }
        }

        internal event EventHandler SelectedValueChanged;

        internal void Popup()
        {
            if (this._anchorControl != null) Utilities.LocateDropDown(this, this._anchorControl, true);
            
            this.Show();
            this.TopMost = true;
        }

        private void Calendar_DateSelected(object sender, System.Windows.Forms.DateRangeEventArgs e)
        {
            this._value = Calendar.SelectionStart;
            SelectedValueChanged(this, new System.EventArgs());
        }

        private void CalendarPopup_Deactivate(object sender, System.EventArgs e)
        {
            this.Hide();
        }

        //  for changes to calendar size we don't control, like when week numbers are added
        private void Calendar_SizeChanged(object sender, System.EventArgs e)
        {
            SizeMe();
        }

        private void SizeMe()
        {
            this.Border.Size = Size.Add(this.Calendar.Size, borderSize);
            this.Size = Size.Add(this.Calendar.Size, borderSize);
        }

        protected override void OnLoad(System.EventArgs e)
        {
            base.OnLoad(e);
            SizeMe();
        }

        protected override void OnFontChanged(System.EventArgs e)
        {
            base.OnFontChanged(e);
            SizeMe();
        }

        protected override void OnShown(System.EventArgs e)
        {
            base.OnShown(e);
            if (this._anchorControl != null) Utilities.LocateDropDown(this, this._anchorControl, true);
        }

        protected override void OnForeColorChanged(System.EventArgs e)
        {
            base.OnForeColorChanged(e);
            this.Calendar.ForeColor = base.ForeColor;
        }
    }
}

File: Library/activiser.Library.DateTimePicker.Desktop/DatePickerPopup.vb
Imports System.ComponentModel

Friend Class DatePickerPopup
    Friend Event SelectedValueChanged As EventHandler ' Of DatePickerValueChangedEventArgs)

    Private Shared ReadOnly borderSize As New Size(2, 2)
    Private _value As DateTime
    Private _anchorControl As Control

#Region "Friend Properties"
    Friend Property Value() As DateTime
        Get
            Return _value
        End Get
        Set(ByVal value As DateTime)
            _value = value
            Me.Calendar.SelectionStart = value
        End Set
    End Property

    Friend Property AnchorControl() As Control
        Get
            Return _anchorControl
        End Get
        Set(ByVal value As Control)
            _anchorControl = value
            If value IsNot Nothing Then
                locateDropDown(Me, Me._anchorControl, True)
            End If
        End Set
    End Property
#End Region

#Region "Friend Methods"
    Friend Sub Popup()
        If Me._anchorControl IsNot Nothing Then
            locateDropDown(Me, Me._anchorControl, True)
        End If
        Me.Show()
        Me.TopMost = True
    End Sub
#End Region

#Region "Component Event Handlers"
    Private Sub Calendar_DateSelected(ByVal sender As System.Object, ByVal e As System.Windows.Forms.DateRangeEventArgs) Handles Calendar.DateSelected
        Me._value = Calendar.SelectionStart
        RaiseEvent SelectedValueChanged(Me, New System.EventArgs)
        'RaiseEvent ValueChanged(Me, New DatePickerValueChangedEventArgs(Me.Value))
    End Sub

    Private Sub CalendarPopup_Deactivate(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Deactivate
        Me.Hide()
    End Sub

    ' for changes to calendar size we don't control, like when week numbers are added
    Private Sub Calendar_SizeChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Calendar.SizeChanged
        SizeMe()
    End Sub

#End Region

#Region "Private Methods"
    Private Sub SizeMe()
        Me.Border.Size = Drawing.Size.Add(Me.Calendar.Size, borderSize)
        Me.Size = Drawing.Size.Add(Me.Calendar.Size, borderSize)
    End Sub
#End Region

#Region "Overridden Event Handlers"
    Protected Overrides Sub OnLoad(ByVal e As System.EventArgs)
        MyBase.OnLoad(e)
        SizeMe()
    End Sub

    Protected Overrides Sub OnFontChanged(ByVal e As System.EventArgs)
        MyBase.OnFontChanged(e)
        SizeMe()
    End Sub

    Protected Overrides Sub OnShown(ByVal e As System.EventArgs)
        MyBase.OnShown(e)
        If Me._anchorControl IsNot Nothing Then
            locateDropDown(Me, Me._anchorControl, True)
        End If
    End Sub

    Protected Overrides Sub OnForeColorChanged(ByVal e As System.EventArgs)
        MyBase.OnForeColorChanged(e)
        Me.Calendar.ForeColor = MyBase.ForeColor
    End Sub
#End Region

    '<DefaultValue(GetType(SystemColors), "WindowText")> _
    'Public Overrides Property ForeColor() As System.Drawing.Color
    '    Get
    '        Return MyBase.ForeColor
    '    End Get
    '    Set(ByVal value As System.Drawing.Color)
    '        MyBase.ForeColor = value
    '    End Set
    'End Property
End Class